MPO time evolution

jesse-osborne opened the pull request:
TODO:
- Construction of higher-order evolution MPOs.
- Apply MPO and conjugate transpose to make evolution closer to unitary.
- Truncation for mp-ievolve-mpo.
- mp-ibc-evolve-mpo?

jesse-osborne committed:
Initial commit for MPO time evolution

ianmccul committed:
Implement UnregularizeBasis12() for DiagonalOperator

jesse-osborne committed:
Add the optimal first-order evolution MPO

jesse-osborne committed:
Add mp-evolve-mpo

ianmccul committed:
Fix implementation of UnregularizeBasis12() for RealDiagonalOperator.

The first implementation was bogus - forgot to copy the appropriate range of the matrix.  The resulting matrix would have had a dimension that was far too big, with multiple copies of elements.  Not sure why this didn't cause assert failures elsewhere, but since the matrices were too big there were at least no memory access violations.

ianmccul committed:
Remove spurious TRACE() calls

jesse-osborne committed:
mp-evolve-mpo: Normalise wavefunction

jesse-osborne committed:
Merge branch 'main' into mpo-evolution

jesse-osborne committed:
Bug fix for constructing the evolution MPO with empty blocks

jesse-osborne commented:
At the moment, the code can’t handle MPOs where the left or right basis is dimension 2, e.g. $`\begin{pmatrix}I&S^z\\0&I\end{pmatrix}`$. If the entire MPO is of this form, we can just apply $`\mathrm{e}^{\mathrm{i}tS^z}`$ to each site, but there may be situations where just one bond in the MPO has dimension 2.

ianmccul commented:
What exactly is the problem with that?  The transformation of the MPO implies that a 2x2 MPO that is 
```math
\begin{pmatrix} I & A \\ 0 & I\end{pmatrix}
```
 transforms into the 1x1 MPO
```math
\begin{pmatrix}
\exp(\tau A)
\end{pmatrix}
```

jesse-osborne commented:
>What exactly is the problem with that?

There is no problem in principle, but the current implementation assumes the MPO to have the form $`\begin{pmatrix}I&C&D\\0&A&B\\0&0&I\end{pmatrix}`$, and fails when $A$, $B$, or $C$ are empty. Incidentally, since the top-left term in the evolution MPO only depends on $D$, it should always be possible to get $\exp(\tau D)$ exactly, but in Van Damme’s paper, they use a Taylor series of $\exp(\tau D)$ only expanded to order $n+1$. Is this something they overlooked, or it somehow worse to use $\exp(\tau D)$ instead of the Taylor series?

ianmccul commented:
Oh OK.  It shouldn't be hard to fix that I guess.  For $m \times n$ MPO, $C$ is always $1 \times (n-2)$.  Matrices with some dimension of zero are actually well defined in some cases, eg earlier this year the SVD of an $m \times 0$ or $0 \times n$ matrix works properly -- if you use the version that has the number of singular vectors as $min(m,n)$ then you get 0 singular vectors, but if you use the other variants then you get $m$ or $n$ random vectors on whichever side is non-zero.

I don't think there is any harm in writing the time evolution of $D$ as $\exp(\tau D)$, but it probably doesn't help much either, since it doesn't change the overall order of the evolution (as long as there is at least one non-zero $C$ or $B$ somewhere).  I guess it does reduce the prefactor of the error though, even if just a bit.  (and if the MPO really is 2x2 then it is exact -- I'm not sure if that is what you really want if you have that MPO but specifically request an order $k$ expansion.... perhaps it should be an option?  I'd be happy for it to default to calculating evolution of $D$ exactly.)

ianmccul commented:
What would help here I guess is a function that extracts a range from an MPO.  I *think* that you can access a subset of a sparse matrix with the notation `M(1,range(1,M.size1()-1))` etc, but it might have bugs (probably never tested), and I guess that isn't exposed in the OperatorComponent either.

Ok, I just checked -- the `SparseMatrix::operator()` isn't implemented for ranges or slicing.  And probably not worth adding it on the main branch, since I want to replace the LinearAlgebra library with the new stuff on the cuda branch one day ... but I reckon it would be worth writing some helper functions, say `extract_range` and `set_range` (or similar names) to return an MPO that is a sub-matrix, and set a sub-matrix of an MPO.  It could take a pair of `LinearAlgebra::range()`, or an int as meaning a length 1 range.  They could eventually be updated to use the MPO operator() (or maybe operator[]!)

ianmccul committed:
Merge remote-tracking branch 'origin/main' into mpo-evolution

ianmccul commented:
So what is the problem?  It works for me making a spinchain and evolving it with sum_unit(Sz(0)).  I only tried finite system so far, I haven't tried mp-ievolve-mpo yet.

There is a weird problem though, I notice by accident that 
```
spinchain-u1 -o lat
mp-random -l lat -u 4 -q 1 -o psi
mp-evolve-mpo -H lat:"sum_unit(Sz(0))" -t 0.01 -n 3 -o psi -w psi
mp-overlap psi psi.t0.01 --rad
```
produces -0.010000041666354 in optimized mode, but produces 3.1315926119234 in debug mode.  That is really weird.  overlap with psi.t0.01 and psi.t0.02 gives the expected result in both cases.

ianmccul commented:
I figured out what is going on; the normalization of finite wavefunctions is a bit messed up, but I think it is fixable. mp-evolve-mpo was throwing away the amplitude (which in debug mode can include a phase, since it randomizes the phase of singular values), but the FiniteWavefunctionLeft constructor wasn't using the phase properly anyway.  I'll add a commit once I get a chance to test it more.

ianmccul committed:
mp-evolve-mpo: keep track of the amplitude

ianmccul committed:
Add argument list versions of tensor_row_sum and tensor_col_sum, add tensor_join

ianmccul committed:
In FiniteWavefunctionLeft::ConstructFromRightOrthogonal(), it is more efficient to use OrthogonalizeBasis2() versus ExpandBasis2() followed by matrix SVD

ianmccul committed:
Use tensor_join() as a way to join components into a larger MPO

ianmccul committed:
Use tensor_join() in mp-evolve-mpo

ianmccul committed:
Use the exact exponential for the D component of the MPO exponential

ianmccul committed:
Added MPO exponential to the operator parser

ianmccul committed:
expmpo() function for approximate MPO evolution using https://arxiv.org/abs/2302.14181

Implements the 1st order and 2nd order schemes, improved by using the Dyson series to incorporate the 'D' terms (local terms) to arbitrary order.

mp-evolve-mpo uses this scheme; mp-ievolve-mpo still needs to be updated.

Also needs adding, allow modifying the DysonOrder parameter for expmpo(), currently the default is order 6.

ianmccul committed:
Bug fix: the Dyson series was finishing at DysonOrder-1, rather than DysonOrder.

Arguably, the order of the dyson series should be reduced for terms that are already higher order, so we cut off the series at a consistent order in Tau.  But maybe not worthwhile, we cannot prevent higher order terms from appearing in the MPO anyway, so that would just be a performance optimization for big MPO's

ianmccul commented:
[notes-on-expmpo.pdf](https://github.com/user-attachments/files/17469044/notes-on-expmpo.pdf)

My notes on the MPO exponential.  The current algorithm implements equation (18)

ianmccul committed:
Fix factorial denominator in the Dyson series

ianmccul committed:
Improved comments in expmpo.cpp
